Add Model Scale Support

JosiahJack opened the pull request:
Add support for models to use `scale` field.  Used to scale up or down misc_model decorations such as statues or trees to add variety and emphasis.

This is purely scaling the model drawing.  May need to scale the mins/maxs to prevent the entity becoming invisible when not looking at its origin (see bug in QSS for example).

andrei-drexler commented:
Thanks! I'm out of town at the moment, I'll take a closer look in a few days when I get back. Glancing over the changes, though, there are a few things that I noticed:
- I think there are a couple of hardcoded sizes that also need to be updated to account for the extra scale byte (the space reserved in the signon buffer in PF_makestatic should be 34 bytes instead of 33, and the overflow threshold in SV_WriteEntitiesToClient should be 40 bytes instead of 39)
- the extra matrix operation in R_DrawAliasModel_Real is effectively a no-op for almost all models, it might be a good idea to only apply it when scale is different from 1. It probably won't make any difference performance-wise in most cases, but it might help in some extreme maps with tons of models
- QSS had an extra scaling factor of 16 when converting from float to byte, but I can't find that here - doesn't that mean that only integral scale factors are supported now? A test map with a few models using fractional scales from say 0.5 to 2.5 would come in really handy
- I'm not sure the comments make the code any easier to read/understand; IMO it's one more thing to parse that doesn't add any extra info other than attribution, but we already have source control for that

mhQuake commented:
This also needs to interact with bounding boxes and clipping hulls. It seems such a niche feature that I'd question if the complexity is worth it.

JosiahJack commented:
Re andrei-drexler:
> * I think there are a couple of hardcoded sizes that also need to be updated...
Ah yes, this is why we have these.  Fixed.
> * the extra matrix op... might be a good idea to only apply it when scale is different from 1...
I almost did this initially.  Done now.  Performance is key for ironwail and the main impetus for my recent attention and attraction to it.
> * QSS had an extra scaling factor of 16...
Fixed to allow for fractional scaling up and down with a value of 16u = 1.0f.
> * I'm not sure the comments...
Removed my attribution comments, definitely fluff.  I was attempting to match the ancient style I've seen sprinkled here and there and an old style guide, but not necessary at all to plaster my handle everywhere.

Re: mhQuake
> * This also needs to interact with bounding boxes and clipping hulls...
Potentially, but most use-cases would not benefit from this at all and may provide more confusion to modders.  Also, it doesn't appear that other engines that support scale bother with this at all, at least QSS doesn't.  Best to have this be purely visual and rely on the modder to apply setsize appropriately and rely on the mapper to clip properly.  Typically I've seen this used on trees and statues most often, neither of which are box-shaped at all.

mhQuake commented:
If it's only ever used for scene decoration objects, which you would makestatic, then yes, clipping hulls aren't a problem.

Bounding boxes remain important.

This is the old GLQuake bug where certain models would disappear at the edges of the viewport. For static entities the bounding box is actually used in two different places: most obviously for frustum culling, but also for generating the efrags list that determines which leafs the static entity is in. If you scale up but don't update the bounding box to match, you will end up not drawing objects that should be visible. 

The other side of this is also true. If you scale down but don't update the bounding box to match, then you could end up doing a lot of unnecessary setup and drawing for objects which are actually outside of the players view.

It could be argued that most scaling will be in very small ranges, which would reduce the impact of this, but that would be relying on something that's out of your control.

JosiahJack commented:
Updated gl_rmain.c to apply scale when appropriate to the entity bounds.  Confirmed that without this change, certain scaled models would pop in and out depending on which vis leafs were visible and that once added, this no longer occurs.

JosiahJack commented:
Applied fixes from sister PR's, squashed.

Shpoike commented:
regarding hulls, this does basically require mods adjust the mins_z values to ensure the entity doesn't leave/penetrate the floor.
for small changes the existing hulls never quite matched the model in the first place. but yeah, don't change mins_x/y (nor maxs_x/y) away from a usable hull size... and mods that ensure these best-practises will be unable to benefit from bsp formats that DO allow arbitrary sizes... but at least they won't misbehave too much. so yeah, gamecode's problem.

JosiahJack commented:
> ...For static entities the bounding box is actually used in two different places...also for generating the efrags list...

Added the scaling to R_AddEfrags.

JosiahJack commented:
I haven't forgotten about this.  I'm still testing different things here to address the issues brought up and discovered in the QS PR.  Spending some time with family and be back at this next week.

Appreciate your patience leaving this open for now.

andrei-drexler commented:
Hey! With the final code getting merged in QS and included in a public release, I figured at this point it's best to just sync with upstream and handle the conflicts. As of 13c299080d32dc470ec91f4bd7711d684fdd9d88 the feature should be functional. Thanks!
